src/tools/rbd_mirror/PoolWatcher.cc

```diff
@@ -23,123 +27,529 @@ using std::list;
+    return;
+  }
+
+  // first time registering or the watch failed
+  dout(5) << dendl;
+  m_async_op_tracker.start_op();
+
+  Context *ctx = create_context_callback<
+    PoolWatcher, &PoolWatcher<I>::handle_register_watcher>(this);
+  m_mirroring_watcher->register_watch(ctx);
+}
+
+template <typename I>
+void PoolWatcher<I>::handle_register_watcher(int r) {
+  dout(5) << "r=" << r << dendl;
+
+  {
+    Mutex::Locker locker(m_lock);
+    assert(m_image_ids_invalid);
+    assert(m_refresh_in_progress);
+    if (r < 0) {
+      m_refresh_in_progress = false;
+    }
+  }
+
+  Context *on_init_finish = nullptr;
   if (r >= 0) {
-    m_images = std::move(image_ids);
+    refresh_images();
   } else if (r == -EBLACKLISTED) {
-    derr << "blacklisted during image refresh" << dendl;
+    dout(0) << "detected client is blacklisted" << dendl;
+
+    Mutex::Locker locker(m_lock);
     m_blacklisted = true;
+    std::swap(on_init_finish, m_on_init_finish);
+  } else if (r == -ENOENT) {
+    dout(5) << "mirroring directory does not exist" << dendl;
+    schedule_refresh_images(30);
+  } else {
+    derr << "unexpected error registering mirroring directory watch: "
+         << cpp_strerror(r) << dendl;
+    schedule_refresh_images(10);
   }
 
-  if (!m_stopping && reschedule) {
-    FunctionContext *ctx = new FunctionContext(
-      boost::bind(&PoolWatcher::refresh_images, this, true));
-    m_timer.add_event_after(m_interval, ctx);
+  m_async_op_tracker.finish_op();
+  if (on_init_finish != nullptr) {
+    on_init_finish->complete(r);
+  }
+}
+
+template <typename I>
+void PoolWatcher<I>::unregister_watcher() {
+  dout(5) << dendl;
+
+  m_async_op_tracker.start_op();
+  Context *ctx = new FunctionContext([this](int r) {
+      dout(5) << "unregister_watcher: r=" << r << dendl;
+      if (r < 0) {
+        derr << "error unregistering watcher for "
+             << m_mirroring_watcher->get_oid() << " object: " << cpp_strerror(r)
+             << dendl;
+      }
+      m_async_op_tracker.finish_op();
+    });
+
+  m_mirroring_watcher->unregister_watch(ctx);
+}
+
+template <typename I>
+void PoolWatcher<I>::refresh_images() {
+  dout(5) << dendl;
+
+  {
+    Mutex::Locker locker(m_lock);
+    assert(m_image_ids_invalid);
+    assert(m_refresh_in_progress);
+
+    // clear all pending notification events since we need to perform
+    // a full image list refresh
+    m_pending_added_image_ids.clear();
+    m_pending_removed_image_ids.clear();
+    if (!m_updated_images.empty()) {
+      auto it = m_updated_images.begin();
+      it->invalid = true;
+
+      // only have a single in-flight request -- remove the rest
+      ++it;
+      while (it != m_updated_images.end()) {
+        m_id_to_updated_images.erase({it->global_image_id,
+                                      it->remote_image_id});
+        it = m_updated_images.erase(it);
+      }
+    }
+    for (auto &updated_image : m_updated_images) {
+      updated_image.invalid = true;
+    }
```

Comment:
Do we need this `for` loop, taking that after the code above we should have a single (invalid) in-flight request?